custom_components/detailed_hello_world/sensor.py

```diff
@@ -0,0 +1,157 @@
+        # Sensors should also register callbacks to HA when their state changes
+        self._roller.registercallback(self.async_write_ha_state)
+
+
+class BatterySensor(SensorBase):
+    """Representation of a Sensor."""
+
+    # The class of this device. Note the value should come from the homeassistant.const
+    # module. More information on the available devices classes can be seen here:
+    # https://developers.home-assistant.io/docs/core/entity/sensor
+    device_class = DEVICE_CLASS_BATTERY
+
+    def __init__(self, roller):
+        """Initialize the sensor."""
+        super().__init__(roller)
+        self._state = random.randint(0, 100)
+
+    # As per the sensor, this must be a unique value within this domain. This is done
+    # by using the device ID, and appending "_battery"
+    @property
+    def unique_id(self):
+        """Return Unique ID string."""
+        return f"{self._roller.roller_id}_battery"
+
+    # This property can return additional metadata about this device. Here it's
+    # returning the voltage of the battery. The actual percentage is returned in
+    # the state property below. These values are displayed in the entity details
+    # screen at the bottom below the history graph.
+    # A nummber of defined attributes are available, see the homeassistant.const module
+    # for constants starting with ATTR_*.
+    # Again, if these values change, the async_write_ha_state method should be called.
+    # in this implementation, these values are assumed to be static.
+    # Note this functionality to display addition data on an entity appears to be
+    # exclusive to sensors. This information is not shown in the UI for a cover.
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        attr = {}
+        attr[ATTR_VOLTAGE] = self._roller.battery_voltage
+        return attr
+
+    # The value of this sensor. As this is a DEVICE_CLASS_BATTERY, this value must be
+    # the battery level as a percentage (between 0 and 100)
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._roller.battery_level
+
+    # The unit of measurement for this entity. As it's a DEVICE_CLASS_BATTERY, this
+    # should be UNIT_PERCENTAGE. A number of units are supported by HA, for some
+    # examples, see:
+    # https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return UNIT_PERCENTAGE
+
+    # The same of this entity, as displayed in the entity UI.
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self._roller.name} Battery"
+
+
+# This is another sensor, but more simple compared to the battery above. See the
+# comments above for how each field works.
+class SignalSensor(SensorBase):
```

Comment:
Signal sensors are really the worst. Let's not put them in the example. They are super spammy. If included, make sure they are disabled by default for the entity registry.

Reply:
Probably still a bit spammy in a real world example, but I have changed this to illuminance (trying to have something that's at least sort of real). I'm keen to include more than one example as I ran into a lot of issues getting it going myself if that makes sense? Obviously it can be anything really :)